chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala

```diff
@@ -370,7 +370,15 @@ abstract class Record extends Aggregate {
     elements.toIndexedSeq.reverse.map(e => eltPort(e._2)).mkString("{", ", ", "}")
   }
 
-  private[chisel3] lazy val flatten = elements.toIndexedSeq.flatMap(_._2.flatten)
+  private[core] override def typeEquivalent(that: Data): Boolean = that match {
+    case that: Record =>
+      this.getClass == that.getClass &&
```

Comment:
For Mux they kind of have to be the same class, because you need to pick one as the return type.  The least arbitrary thing to do is to pick one but require it be the same as the other.

For connections, there is a debate to be had, but relaxing it later is always easier than imposing a new restriction.